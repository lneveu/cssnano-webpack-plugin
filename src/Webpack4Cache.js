import os from 'os';

import cacache from 'cacache';
import findCacheDir from 'find-cache-dir';
import serialize from 'serialize-javascript';

export default class Webpack4Cache {
  constructor(compilation, options, weakCache) {
    this.cache =
      options.cache === true
        ? Webpack4Cache.getCacheDirectory()
        : options.cache;
    this.weakCache = weakCache;
  }

  static getCacheDirectory() {
    return (
      findCacheDir({ name: 'css-minimizer-webpack-plugin' }) || os.tmpdir()
    );
  }

  async get(cacheData, sources) {
    const weakOutput = this.weakCache.get(cacheData.assetSource);

    if (weakOutput) {
      return weakOutput;
    }

    if (!this.cache) {
      // eslint-disable-next-line no-undefined
      return undefined;
    }

    // eslint-disable-next-line no-param-reassign
    cacheData.cacheIdent =
      cacheData.cacheIdent || serialize(cacheData.cacheKeys);

    let cachedResult;

    try {
      cachedResult = await cacache.get(this.cache, cacheData.cacheIdent);
    } catch (ignoreError) {
      // eslint-disable-next-line no-undefined
      return undefined;
    }

    cachedResult = JSON.parse(cachedResult.data);

    const { css, map, input, assetName, inputSourceMap } = cachedResult;

    if (map) {
      cachedResult.source = new sources.SourceMapSource(
        css,
        assetName,
        map,
        input,
        inputSourceMap,
        true
      );
    } else {
      cachedResult.source = new sources.RawSource(css);
    }

    return cachedResult;
  }

  async store(cacheData) {
    if (!this.weakCache.has(cacheData.assetSource)) {
      this.weakCache.set(cacheData.assetSource, cacheData);
    }

    if (!this.cache) {
      // eslint-disable-next-line no-undefined
      return undefined;
    }

    const {
      cacheIdent,
      css,
      assetName,
      map,
      input,
      inputSourceMap,
    } = cacheData;

    return cacache.put(
      this.cache,
      cacheIdent,
      JSON.stringify({
        assetName,
        css,
        map,
        input,
        inputSourceMap,
      })
    );
  }
}
