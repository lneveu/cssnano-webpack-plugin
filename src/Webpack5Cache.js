export default class Cache {
  // eslint-disable-next-line no-unused-vars
  constructor(compilation) {
    this.cache = compilation.getCache('CssMinimizerWebpackPlugin');
  }

  async get(task) {
    // eslint-disable-next-line no-param-reassign
    task.eTag = task.eTag || this.cache.getLazyHashedEtag(task.assetSource);

    return this.cache.getPromise(task.assetName, task.eTag);
  }

  async store(task) {
    return this.cache.storePromise(task.assetName, task.eTag, task.output);
  }
}
