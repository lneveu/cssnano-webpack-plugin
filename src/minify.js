const cssnano = require('cssnano');

const minify = async (options) => {
  const {
    input,
    postcssOptions,
    cssnanoOptions,
    map,
    inputSourceMap,
    minify: minifyFn,
  } = options;

  if (minifyFn) {
    return minifyFn({ input, postcssOptions, cssnanoOptions }, inputSourceMap);
  }

  if (inputSourceMap) {
    postcssOptions.map = { prev: inputSourceMap, ...map };
  }

  const result = await cssnano.process(input, postcssOptions, cssnanoOptions);

  return {
    css: result.css,
    map: result.map,
    error: result.error,
    warnings: result.warnings(),
  };
};

async function transform(options) {
  // 'use strict' => this === undefined (Clean Scope)
  // Safer for possible security issues, albeit not critical at all here
  // eslint-disable-next-line no-new-func, no-param-reassign
  options = new Function(
    'exports',
    'require',
    'module',
    '__filename',
    '__dirname',
    `'use strict'\nreturn ${options}`
  )(exports, require, module, __filename, __dirname);

  const result = await minify(options);

  return result;
}

module.exports.minify = minify;
module.exports.transform = transform;
