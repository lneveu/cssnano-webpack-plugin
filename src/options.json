{
  "definitions": {
    "Rule": {
      "description": "Filtering rule as regex or string.",
      "anyOf": [
        {
          "instanceof": "RegExp",
          "tsType": "RegExp"
        },
        {
          "type": "string",
          "minLength": 1
        }
      ]
    },
    "Rules": {
      "description": "Filtering rules.",
      "anyOf": [
        {
          "type": "array",
          "items": {
            "description": "A rule condition.",
            "oneOf": [
              {
                "$ref": "#/definitions/Rule"
              }
            ]
          }
        },
        {
          "$ref": "#/definitions/Rule"
        }
      ]
    }
  },
  "title": "CssnanoWebpackPluginOptions",
  "type": "object",
  "properties": {
    "test": {
      "description": "Include all modules that pass test assertion.",
      "oneOf": [
        {
          "$ref": "#/definitions/Rules"
        }
      ]
    },
    "include": {
      "description": "Include all modules matching any of these conditions.",
      "oneOf": [
        {
          "$ref": "#/definitions/Rules"
        }
      ]
    },
    "exclude": {
      "description": "Exclude all modules matching any of these conditions.",
      "oneOf": [
        {
          "$ref": "#/definitions/Rules"
        }
      ]
    },
    "sourceMap": {
      "description": "Enables/Disables generation of source maps.",
      "anyOf": [
        {
          "description": "Options for source map.",
          "additionalProperties": true,
          "type": "object"
        },
        {
          "type": "boolean"
        }
      ]
    },
    "cssnanoOptions": {
      "description": "Options for `cssnanoOptions`.",
      "additionalProperties": true,
      "type": "object"
    }
  },
  "additionalProperties": false
}
