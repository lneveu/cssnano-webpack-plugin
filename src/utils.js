/* istanbul ignore next */
async function cssnanoMinify(
  data,
  inputSourceMap,
  minimizerOptions = { preset: 'default' }
) {
  const [[name, input]] = Object.entries(data);
  const postcssOptions = {
    to: name,
    from: name,
    ...minimizerOptions.processorOptions,
  };

  if (typeof postcssOptions.parser === 'string') {
    try {
      postcssOptions.parser = await load(postcssOptions.parser);
    } catch (error) {
      throw new Error(
        `Loading PostCSS "${postcssOptions.parser}" parser failed: ${error.message}\n\n(@${name})`
      );
    }
  }

  if (typeof postcssOptions.stringifier === 'string') {
    try {
      postcssOptions.stringifier = await load(postcssOptions.stringifier);
    } catch (error) {
      throw new Error(
        `Loading PostCSS "${postcssOptions.stringifier}" stringifier failed: ${error.message}\n\n(@${name})`
      );
    }
  }

  if (typeof postcssOptions.syntax === 'string') {
    try {
      postcssOptions.syntax = await load(postcssOptions.syntax);
    } catch (error) {
      throw new Error(
        `Loading PostCSS "${postcssOptions.syntax}" syntax failed: ${error.message}\n\n(@${name})`
      );
    }
  }

  if (inputSourceMap) {
    postcssOptions.map = {
      annotation: false,
      prev: inputSourceMap,
    };
  }

  // eslint-disable-next-line global-require
  const cssnano = require('cssnano');
  const result = await cssnano.process(input, postcssOptions, minimizerOptions);

  return {
    code: result.css,
    map: result.map && result.map.toString(),
    warnings: result.warnings().map(String),
  };

  async function load(module) {
    let exports;

    try {
      // eslint-disable-next-line import/no-dynamic-require, global-require
      exports = require(module);

      return exports;
    } catch (requireError) {
      let importESM;

      try {
        // eslint-disable-next-line no-new-func
        importESM = new Function('id', 'return import(id);');
      } catch (e) {
        importESM = null;
      }

      if (requireError.code === 'ERR_REQUIRE_ESM' && importESM) {
        exports = await importESM(module);

        return exports.default;
      }

      throw requireError;
    }
  }
}

// eslint-disable-next-line import/prefer-default-export
export { cssnanoMinify };
