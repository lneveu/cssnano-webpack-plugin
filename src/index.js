import os from 'os';
import crypto from 'crypto';

import { SourceMapConsumer } from 'source-map';
import RequestShortener from 'webpack/lib/RequestShortener';
import webpack, {
  ModuleFilenameHelpers,
  SourceMapDevToolPlugin,
  version as webpackVersion,
} from 'webpack';
import validateOptions from 'schema-utils';
import serialize from 'serialize-javascript';
import CssMinimizerPackageJson from 'cssnano/package.json';
import pLimit from 'p-limit';
import Worker from 'jest-worker';

import schema from './options.json';

import { minify as minifyFn } from './minify';

const warningRegex = /\s.+:+([0-9]+):+([0-9]+)/;

// webpack 5 exposes the sources property to ensure the right version of webpack-sources is used
const { SourceMapSource, RawSource } =
  // eslint-disable-next-line global-require
  webpack.sources || require('webpack-sources');

class CssMinimizerPlugin {
  constructor(options = {}) {
    validateOptions(schema, options, {
      name: 'Css Minimizer Plugin',
      baseDataPath: 'options',
    });

    const {
      minify,
      minimizerOptions = {
        preset: 'default',
      },
      test = /\.css(\?.*)?$/i,
      warningsFilter = () => true,
      sourceMap = false,
      cache = true,
      cacheKeys = (defaultCacheKeys) => defaultCacheKeys,
      parallel = true,
      include,
      exclude,
    } = options;

    this.options = {
      test,
      warningsFilter,
      sourceMap,
      cache,
      cacheKeys,
      parallel,
      include,
      exclude,
      minify,
      minimizerOptions,
    };

    if (this.options.sourceMap === true) {
      this.options.sourceMap = { inline: false };
    }
  }

  static isSourceMap(input) {
    // All required options for `new SourceMapConsumer(...options)`
    // https://github.com/mozilla/source-map#new-sourcemapconsumerrawsourcemap
    return Boolean(
      input &&
        input.version &&
        input.sources &&
        Array.isArray(input.sources) &&
        typeof input.mappings === 'string'
    );
  }

  static buildError(error, file, sourceMap, requestShortener) {
    if (error.line) {
      const original =
        sourceMap &&
        sourceMap.originalPositionFor({
          line: error.line,
          column: error.column,
        });

      if (original && original.source && requestShortener) {
        return new Error(
          `${file} from Css Minimizer Webpack Plugin\n${
            error.message
          } [${requestShortener.shorten(original.source)}:${original.line},${
            original.column
          }][${file}:${error.line},${error.column}]${
            error.stack
              ? `\n${error.stack.split('\n').slice(1).join('\n')}`
              : ''
          }`
        );
      }

      return new Error(
        `${file} from Css Minimizer \n${error.message} [${file}:${error.line},${
          error.column
        }]${
          error.stack ? `\n${error.stack.split('\n').slice(1).join('\n')}` : ''
        }`
      );
    }

    if (error.stack) {
      return new Error(`${file} from Css Minimizer\n${error.stack}`);
    }

    return new Error(`${file} from Css Minimizer\n${error.message}`);
  }

  static buildWarning(
    warning,
    file,
    sourceMap,
    requestShortener,
    warningsFilter
  ) {
    let warningMessage = warning;
    let locationMessage = '';
    let source;

    if (sourceMap) {
      const match = warningRegex.exec(warning);

      if (match) {
        const line = +match[1];
        const column = +match[2];
        const original = sourceMap.originalPositionFor({
          line,
          column,
        });

        if (
          original &&
          original.source &&
          original.source !== file &&
          requestShortener
        ) {
          ({ source } = original);

          warningMessage = `${warningMessage.replace(warningRegex, '')}`;
          locationMessage = `${requestShortener.shorten(original.source)}:${
            original.line
          }:${original.column}`;
        }
      }
    }

    if (warningsFilter && !warningsFilter(warning, file, source)) {
      return null;
    }

    return `Css Minimizer Plugin: ${warningMessage} ${locationMessage}`;
  }

  static getAvailableNumberOfCores(parallel) {
    // In some cases cpus() returns undefined
    // https://github.com/nodejs/node/issues/19022
    const cpus = os.cpus() || { length: 1 };

    return parallel === true
      ? cpus.length - 1
      : Math.min(Number(parallel) || 0, cpus.length - 1);
  }

  // eslint-disable-next-line consistent-return
  static getAsset(compilation, name) {
    // New API
    if (compilation.getAsset) {
      return compilation.getAsset(name);
    }

    if (compilation.assets[name]) {
      return { name, source: compilation.assets[name], info: {} };
    }
  }

  static updateAsset(compilation, name, newSource, assetInfo) {
    // New API
    if (compilation.updateAsset) {
      compilation.updateAsset(name, newSource, assetInfo);
    }

    // eslint-disable-next-line no-param-reassign
    compilation.assets[name] = newSource;
  }

  *getTask(compiler, compilation, assetName) {
    const {
      info: assetInfo,
      source: assetSource,
    } = CssMinimizerPlugin.getAsset(compilation, assetName);

    // Skip double minimize assets from child compilation
    if (assetInfo.minimized) {
      yield false;
    }

    let input;
    let inputSourceMap;

    // TODO refactor after drop webpack@4, webpack@5 always has `sourceAndMap` on sources
    if (this.options.sourceMap && assetSource.sourceAndMap) {
      const { source, map } = assetSource.sourceAndMap();

      input = source;

      if (map) {
        if (CssMinimizerPlugin.isSourceMap(map)) {
          inputSourceMap = map;
        } else {
          inputSourceMap = map;

          compilation.warnings.push(
            new Error(`${assetName} contains invalid source map`)
          );
        }
      }
    } else {
      input = assetSource.source();
      inputSourceMap = null;
    }

    if (Buffer.isBuffer(input)) {
      input = input.toString();
    }

    const task = {
      assetSource,
      assetInfo,
      assetName,
      input,
      inputSourceMap,
      map: this.options.sourceMap,
      minimizerOptions: this.options.minimizerOptions,
      minify: this.options.minify,
    };

    if (CssMinimizerPlugin.isWebpack4()) {
      if (this.options.cache) {
        const defaultCacheKeys = {
          nodeVersion: process.version,
          // eslint-disable-next-line global-require
          'css-minimizer-webpack-plugin': require('../package.json').version,
          cssMinimizer: CssMinimizerPackageJson.version,
          'css-minimizer-webpack-plugin-options': this.options,
          assetName,
          contentHash: crypto.createHash('md4').update(input).digest('hex'),
        };

        task.cacheKeys = this.options.cacheKeys(defaultCacheKeys, assetName);
      }
    }

    yield task;
  }

  // eslint-disable-next-line class-methods-use-this
  async runTasks(compiler, compilation, assetNames, CacheEngine, weakCache) {
    const cache = new CacheEngine(
      compilation,
      {
        cache: this.options.cache,
      },
      weakCache
    );
    const availableNumberOfCores = CssMinimizerPlugin.getAvailableNumberOfCores(
      this.options.parallel
    );

    let concurrency = Infinity;
    let worker;

    if (availableNumberOfCores > 0) {
      // Do not create unnecessary workers when the number of files is less than the available cores, it saves memory
      const numWorkers = Math.min(assetNames.length, availableNumberOfCores);

      concurrency = numWorkers;

      worker = new Worker(require.resolve('./minify'), { numWorkers });

      // https://github.com/facebook/jest/issues/8872#issuecomment-524822081
      const workerStdout = worker.getStdout();

      if (workerStdout) {
        workerStdout.on('data', (chunk) => {
          return process.stdout.write(chunk);
        });
      }

      const workerStderr = worker.getStderr();

      if (workerStderr) {
        workerStderr.on('data', (chunk) => {
          return process.stderr.write(chunk);
        });
      }
    }

    const limit = pLimit(concurrency);
    const scheduledTasks = [];

    for (const assetName of assetNames) {
      scheduledTasks.push(
        limit(async () => {
          const task = this.getTask(compiler, compilation, assetName).next()
            .value;

          if (!task) {
            return Promise.resolve();
          }

          let resultOutput = await cache.get(task, {
            RawSource,
            SourceMapSource,
          });

          const { inputSourceMap } = task;

          let sourceMap;

          if (!resultOutput) {
            try {
              // eslint-disable-next-line no-param-reassign
              resultOutput = worker
                ? await worker.transform(serialize(task))
                : await minifyFn(task);
            } catch (error) {
              task.error = error;
            }

            task.error = task.error || resultOutput.error;

            if (task.error) {
              if (
                inputSourceMap &&
                CssMinimizerPlugin.isSourceMap(inputSourceMap)
              ) {
                sourceMap = new SourceMapConsumer(inputSourceMap);
              }

              compilation.errors.push(
                CssMinimizerPlugin.buildError(
                  task.error,
                  assetName,
                  sourceMap,
                  new RequestShortener(compiler.context)
                )
              );

              return Promise.resolve();
            }

            task.css = resultOutput.css;
            task.map = resultOutput.map;
            task.warnings = resultOutput.warnings;

            const { css: code, map, input } = task;

            if (map) {
              task.source = new SourceMapSource(
                code,
                assetName,
                map,
                input,
                inputSourceMap,
                true
              );
            } else {
              task.source = new RawSource(code);
            }

            await cache.store(task);
          } else {
            task.source = resultOutput.source;
            task.warnings = resultOutput.warnings;
          }

          if (task.warnings && task.warnings.length > 0) {
            if (
              inputSourceMap &&
              CssMinimizerPlugin.isSourceMap(inputSourceMap)
            ) {
              sourceMap = new SourceMapConsumer(inputSourceMap);
            }

            task.warnings.forEach((warning) => {
              const builtWarning = CssMinimizerPlugin.buildWarning(
                warning,
                assetName,
                sourceMap,
                new RequestShortener(compiler.context),
                this.options.warningsFilter
              );

              if (builtWarning) {
                compilation.warnings.push(builtWarning);
              }
            });
          }

          const { source, assetInfo } = task;

          CssMinimizerPlugin.updateAsset(compilation, assetName, source, {
            ...assetInfo,
            minimized: true,
          });

          return Promise.resolve();
        })
      );
    }

    await Promise.all(scheduledTasks);

    if (worker) {
      await worker.end();
    }
  }

  static isWebpack4() {
    return webpackVersion[0] === '4';
  }

  apply(compiler) {
    const pluginName = this.constructor.name;
    const { devtool, plugins } = compiler.options;

    this.options.sourceMap =
      typeof this.options.sourceMap === 'undefined'
        ? (devtool &&
            !devtool.includes('eval') &&
            !devtool.includes('cheap') &&
            (devtool.includes('source-map') ||
              // Todo remove when `webpack@4` support will be dropped
              devtool.includes('sourcemap'))) ||
          (plugins &&
            plugins.some(
              (plugin) =>
                plugin instanceof SourceMapDevToolPlugin &&
                plugin.options &&
                plugin.options.columns
            ))
        : this.options.sourceMap;

    const matchObject = ModuleFilenameHelpers.matchObject.bind(
      // eslint-disable-next-line no-undefined
      undefined,
      this.options
    );

    const weakCache = new WeakMap();

    const optimizeFn = async (compilation, CacheEngine, assets) => {
      const assetNames = Object.keys(
        typeof assets === 'undefined' ? compilation.assets : assets
      ).filter((assetName) => matchObject(assetName));

      if (assetNames.length === 0) {
        return Promise.resolve();
      }

      await this.runTasks(
        compiler,
        compilation,
        assetNames,
        CacheEngine,
        weakCache
      );

      return Promise.resolve();
    };

    compiler.hooks.compilation.tap(pluginName, (compilation) => {
      if (this.options.sourceMap) {
        compilation.hooks.buildModule.tap(pluginName, (moduleArg) => {
          // to get detailed location info about errors
          // eslint-disable-next-line no-param-reassign
          moduleArg.useSourceMap = true;
        });
      }

      if (CssMinimizerPlugin.isWebpack4()) {
        // eslint-disable-next-line global-require
        const CacheEngine = require('./Webpack4Cache').default;

        compilation.hooks.optimizeChunkAssets.tapPromise(pluginName, () =>
          // eslint-disable-next-line no-undefined
          optimizeFn(compilation, CacheEngine, undefined, weakCache)
        );
      } else {
        if (this.options.sourceMap) {
          compilation.hooks.buildModule.tap(pluginName, (moduleArg) => {
            // to get detailed location info about errors
            // eslint-disable-next-line no-param-reassign
            moduleArg.useSourceMap = true;
          });
        }

        // eslint-disable-next-line global-require
        const CacheEngine = require('./Webpack5Cache').default;

        // eslint-disable-next-line global-require
        const Compilation = require('webpack/lib/Compilation');

        compilation.hooks.processAssets.tapPromise(
          {
            name: pluginName,
            stage: Compilation.PROCESS_ASSETS_STAGE_OPTIMIZE_SIZE,
          },
          (assets) => optimizeFn(compilation, CacheEngine, assets)
        );

        compilation.hooks.statsPrinter.tap(pluginName, (stats) => {
          stats.hooks.print
            .for('asset.info.minimized')
            .tap(
              'css-minimizer-webpack-plugin',
              (minimized, { green, formatFlag }) =>
                // eslint-disable-next-line no-undefined
                minimized ? green(formatFlag('minimized')) : undefined
            );
        });
      }
    });
  }
}

export default CssMinimizerPlugin;
