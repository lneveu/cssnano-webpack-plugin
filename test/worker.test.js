import serialize from 'serialize-javascript';

import { transform } from '../src/minify';

import { normalizeErrors } from './helpers';

describe('worker', () => {
  it('should minify css', async () => {
    const options = {
      input: '.foo{color:red;}\n.bar{color:coral;}',
      postcssOptions: {
        from: 'entry.css',
        to: 'entry.css',
        map: {
          prev: {
            version: 3,
            sources: ['foo.css', 'bar.css'],
            names: [],
            mappings: 'AAAA,KAAK,iBAAiB,KAAK,UAAU,OAAO',
            file: 'x',
            sourcesContent: ['.foo{color:red;}', '.bar{color:coral;}'],
          },
          inline: false,
        },
      },
      cssnanoOptions: { discardComments: false },
    };
    const { css, map } = await transform(serialize(options));

    expect(css).toMatchSnapshot('css');
    expect(map).toMatchSnapshot('map');
  });

  it('should work inputSourceMap as prev', async () => {
    const options = {
      input: '.foo{color:red;}\n.bar{color:coral;}',
      postcssOptions: {
        from: 'entry.css',
        to: 'entry.css',
        map: {
          prev: {
            version: 3,
            sources: ['foo.css', 'bar.css'],
            names: [],
            mappings: 'AAAA,KAAK,iBAAiB,KAAK,UAAU,OAAO',
            file: 'x',
            sourcesContent: ['.foo{color:red;}', '.bar{color:coral;}'],
          },
          inline: false,
        },
      },
      cssnanoOptions: { discardComments: false },
      inputSourceMap: {
        version: 3,
        sources: ['foo.css', 'bar.css'],
        names: [],
        mappings: 'AAAA,KAAK,iBAAiB,KAAK,UAAU,OAAO',
        file: 'x',
        sourcesContent: ['.foo{color:red;}', '.bar{color:coral;}'],
      },
    };
    const { css, map } = await transform(serialize(options));

    expect(css).toMatchSnapshot('css');
    expect(map).toMatchSnapshot('map');
  });

  it('should work options.minify function', async () => {
    const options = {
      input: '.foo{color:red;}\n.bar{color:coral;}',
      postcssOptions: {
        from: 'entry.css',
        to: 'entry.css',
      },
      cssnanoOptions: { discardComments: false },
      minify: () => {
        return { css: '.minify {};' };
      },
    };
    const { css, map } = await transform(serialize(options));

    expect(css).toMatchSnapshot('css');
    expect(map).toMatchSnapshot('map');
  });

  it('should emit error', async () => {
    const options = {
      input: false,
      postcssOptions: {
        from: 'entry.css',
        to: 'entry.css',
      },
    };

    try {
      await transform(serialize(options));
    } catch (error) {
      const normalizeError = { ...error };

      normalizeError.message = [error.message.split('\n')];

      expect(normalizeErrors(normalizeError.message)).toMatchSnapshot('error');
    }
  });

  it('should emit minimizer error', async () => {
    const options = {
      input: false,
      postcssOptions: {
        from: 'entry.css',
        to: 'entry.css',
      },
      minify: () => {
        return { error: new Error('cssnano error') };
      },
    };

    try {
      await transform(serialize(options));
    } catch (error) {
      const normalizeError = { ...error };

      normalizeError.message = [error.message.split('\n')];

      expect(normalizeErrors(normalizeError.message)).toMatchSnapshot('error');
    }
  });
});
