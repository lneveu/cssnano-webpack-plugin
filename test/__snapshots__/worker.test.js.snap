// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`worker should emit error: error 1`] = `
Array [
  "/entry.css:1:1: Unknown word",
]
`;

exports[`worker should minify css: css 1`] = `".foo{color:red}.bar{color:coral}"`;

exports[`worker should minify css: map 1`] = `"{\\"version\\":3,\\"sources\\":[\\"foo.css\\",\\"entry.css\\"],\\"names\\":[],\\"mappings\\":\\"AAAA,KAAK,SAAA,CCCL,KAAK,WAAY\\",\\"file\\":\\"entry.css\\",\\"sourcesContent\\":[\\".foo{color:red;}\\",\\".foo{color:red;}\\\\n.bar{color:coral;}\\"]}"`;

exports[`worker should work inputSourceMap as prev: css 1`] = `".foo{color:red}.bar{color:coral}"`;

exports[`worker should work inputSourceMap as prev: map 1`] = `"{\\"version\\":3,\\"sources\\":[\\"foo.css\\",\\"entry.css\\"],\\"names\\":[],\\"mappings\\":\\"AAAA,KAAK,SAAA,CCCL,KAAK,WAAY\\",\\"file\\":\\"entry.css\\",\\"sourcesContent\\":[\\".foo{color:red;}\\",\\".foo{color:red;}\\\\n.bar{color:coral;}\\"]}"`;

exports[`worker should work options.minify function: css 1`] = `".minify {};"`;

exports[`worker should work options.minify function: map 1`] = `undefined`;
